src/models/StyledComponent.js

```diff
@@ -50,20 +50,32 @@ export default (ComponentStyle: Function) => {
             this.setState({ theme, generatedClassName })
           })
         } else {
+          const theme = this.props.theme || {}
           const generatedClassName = this.generateAndInjectStyles(
-            this.props.theme || {},
+            theme,
             this.props
           )
-          this.setState({ generatedClassName })
+          this.setState({ theme, generatedClassName })
         }
       }
 
       componentWillReceiveProps(nextProps: any) {
+        if (!nextProps.theme) {
+          return
+        }
+
+        // compare if there is any change
+        if (JSON.stringify(nextProps.theme) === JSON.stringify(this.state.theme)) {
```

Comment:
@mxstbr I don't know if there is need to do other equality check than `===`. Because if we pass the same theme to `generateAndInjectStyles()` we should get the same result right?

I am looking on this https://github.com/substack/node-deep-equal.